src/cookbooks/bigquery/optimization.md

```diff
@@ -1,28 +1,165 @@
+
+For example, consider the following query, which lists the number of rows that are present for each
+`client_id`:
+
+```
+SELECT client_id, COUNT(*)
+FROM telemetry.main
+GROUP BY client_id
+```
+
+During the execution of this query, the BigQuery cluster performs the following steps:
+
+1. Each system reads a different piece of the data and parses the `client_id` for
+   each row. Internally, it then computes the number of rows seen for each `client_id`,
+   _but only for the data that it read_.
+1. Each system is then assigned a set of `client_id`s to aggregate. For example, the first
+   system may be given instructions to get the count of `client1`. It then has to send a request to every other system for the total seen for `client1`. It can then aggregate the total.
+1. If every `client_id` has been aggregated, each system reports to the coordinator
+   the `client_id`s that it was responsible for, as well as the count of rows seen by each.
+   The coordinator is responsible for returning the result of the query to the client,
+   which in this example is STMO.
+
+A similar process occurs on data joins, where different systems are instructed to join on
+different keys. In that case, data from both tables needs to be shuffled to every system.
+
+#### Key takeaways
+
+- Use `LIMIT` for query prototyping to dramatically reduce the volume of data scanned
+  as well as speeding up processing.
+- Use approximate algorithms. Then less data needs to be shuffled because
+  probabilistic data structures can be used instead of the raw data itself.
+- Specify large tables first in a `JOIN` operation. In this case, small tables can be sent to
+  every system to eliminate one data shuffle operation. Note that Spark supports a `broadcast`
+  command explicitly.
+
+### How is the data stored?
+
+The data is stored in columnar format.
+
+#### Traditional Row Stores
+
+Consider a typical CSV file, which represents an example of a row store.
+
+```
+name,age,height
+"Ted",27,6.0
+"Emmanuel",45,5.9
+"Cadence",5,3.5
+```
+
+When this data is stored to disk, you can read an entire record in consecutive order. For example, if
+the first `"` is stored at block 1 on disk, then a sequential scan from 1 assigns the first row of
+data: `"ted",27,6.0`. Keep scanning and you get `\n"Emm`... and so on.
+
+You can use the following query that you can execute very quickly:
+
+```
+SELECT *
+FROM people
+WHERE name == 'Ted'
+```
+
+The database can just scan the first row of data. However, the following is more difficult:
+
+```
+SELECT name
+FROM people
+```
+
+Now the database has to read _all_ of the rows and then select the `name` column, which results in a lot
+more overhead.
+
+#### Columnar Stores
+
+Columnar turns the data sideways. For example, you can make a columnar version of the above data
+and still store it in CSV format:
+
+```
+name,"Ted","Emmanuel","Cadence"
+age,27,45,5
```

Comment:
Side note: my daughter, Cadence, was 5 when I wrote this, she just turned 8 yesterday 😭